### ISSUE TITLE
Save a house every X minutes after an item inside it is moved / added / removed.

### ISSUE THREAD

**Issue by brunotonin** [otland/forgottenserver#3378]
### Explanation of what you want to do that is currently impossible
Save a house every X minutes after an item inside it is moved / added / removed.
This would be very good for cases of crash on the server, the players do not lose the items in their homes.

### Desired functionality
I think the ideal would be that every time an item is changed in the house, schedule to save the items in the house in X minutes, to avoid many interactions with the DB. event of saving the previous house is canceled and another one is created, or the time of the previous one is reset.

**Comment by MillhioreBT** [comment]
It sounds interesting, but I don't know how efficient this would be.

However, it is assumed that the server should not suffer crashes if it is in production, it would be a very specific case and very very rare, and usually it is our own fault, you know some script with infinite loops, some incorrect call to a already  deleted object in C or something like that

but it would be nice to add a function that does this work and you would work this case yourself from a script in the onItemMoved event

**Comment by brunotonin** [comment]
> saveHouse(...)

Really the ideal would be that it does not happen, however many new servers have this problem and when it happens with big servers the impact of this is gigantic and gives a huge headache, because normally the server save happens every 24 hours, and in an eventual crash, it could lose a lot in the houses.

About adding a function` saveHouse(...)`, I think that would solve it because its implementation would be simple and it makes sense to have a function like that inside tfs.

**Comment by brunotonin** [comment]
it would be interesting to do the following test to simulate what would happen if you ran the save in a house that lost its owner.

```
addEvent(funtion(houseId)
    local house = House(houseId)
    house:save()
end, 2000, house:getId())

house:setOwner(0)
```

**Comment by DSpeichert** [comment]
This is request for an anti-pattern. There are advantages but it would also make it easier to exploit any crash if only partial state is saved. Think about cloning items: item is moved to a house, house is saved but player is not (upon a crash) and you have item cloning.

I think this feature, which as pointed out should never actually be useful in production, has potential to do even more harm.

**Comment by Kamenuvol** [comment]
With two players you can exploit crashes too. Player A gives Item to Player B and Player B logs off. Server crash and both players now have the Item.
Using @DSpeichert logic, it should save players and houses only on server save.

**Comment by brunotonin** [comment]
> This is request for an anti-pattern. There are advantages but it would also make it easier to exploit any crash if only partial state is saved. Think about cloning items: item is moved to a house, house is saved but player is not (upon a crash) and you have item cloning.
> 
> I think this feature, which as pointed out should never actually be useful in production, has potential to do even more harm.

The situation in which the clone of the items occurs can already be done today, either using another character (passing the items) or even the house (taking the items from it).
I agree that this change is for specific cases, but the idea would be to reduce the damage caused to the servers and especially for the players.
It is worth mentioning that normally the loss of items in the houses is a much greater damage than the cloning of the items done by a small number of players.

**Comment by DSpeichert** [comment]
You are both right. But we don't exactly develop this project to crash gracefully - we want it to not crash.

I think the ability to NOT save characters when they log out is an equally useful feature that could prevent item cloning.

**Comment by MillhioreBT** [comment]
Although I proposed changes to add such functionality, I will not support the implementation of this feature, It is very dangerous and also somewhat expensive, it would be like the frankenstein's function of TFS XD

**Comment by EPuncker** [comment]
@brunotonin anything else you want to add to the subject or does the #3386 do the job?

**Comment by brunotonin** [comment]
This solution solves the question of saving the houses, @MillhioreBT made an example of how to save the houses when moving an item and added a cooldown to not overload the database, it may be interesting to add an option in the config.lua to activate this script and configure the time it runs.

**Comment by nekiro** [comment]
Closed in #3386
